sockeye/decoder.py

```diff
@@ -745,44 +769,65 @@ def _step(self,
+    def _hidden_mlp(self,
+                    attention_state: attentions.AttentionState,
+                    rnn_output: mx.sym.Symbol,
+                    seq_idx: int) -> mx.sym.Symbol:
+        # hidden: (batch_size, rnn_num_hidden)
+        hidden = mx.sym.FullyConnected(data=mx.sym.concat(rnn_output, attention_state.context, dim=1),
+                                       # use same number of hidden states as RNN
+                                       num_hidden=self.num_hidden,
+                                       weight=self.hidden_w,
+                                       bias=self.hidden_b)
+        if self.config.layer_normalization:
+            hidden = self.hidden_norm.normalize(hidden)
+        if self.config.hidden_dropout > 0:
+            hidden = mx.sym.Dropout(data=hidden, p=self.config.hidden_dropout)
 
-            if self.config.layer_normalization:
-                hidden = self.hidden_norm.normalize(hidden)
+        # hidden: (batch_size, rnn_num_hidden)
+        hidden = mx.sym.Activation(data=hidden, act_type="tanh",
+                                   name="%snext_hidden_t%d" % (self.prefix, seq_idx))
+        return hidden
+
+    def _context_gate(self,
+                      word_vec_prev: mx.sym.Symbol,
+                      rnn_output: mx.sym.Symbol,
+                      attention_state: attentions.AttentionState,
+                      seq_idx: int) -> mx.sym.Symbol:
+        # context: (batch_size, encoder_num_hidden)
+        # gate: (batch_size, rnn_num_hidden)
+        gate = mx.sym.FullyConnected(data=mx.sym.concat(word_vec_prev, rnn_output, attention_state.context, dim=1),
+                                     num_hidden=self.num_hidden, weight=self.gate_w, bias=self.gate_b)
+        gate = mx.sym.Activation(data=gate, act_type="sigmoid",
+                                 name="%sgate_activation_t%d" % (self.prefix, seq_idx))
+
+        # mapped_rnn_output: (batch_size, rnn_num_hidden)
+        mapped_rnn_output = mx.sym.FullyConnected(data=rnn_output,
+                                                  num_hidden=self.num_hidden,
+                                                  weight=self.mapped_rnn_output_w,
+                                                  bias=self.mapped_rnn_output_b,
+                                                  name="%smapped_rnn_output_fc_t%d" % (self.prefix, seq_idx))
+        # mapped_context: (batch_size, rnn_num_hidden)
+        mapped_context = mx.sym.FullyConnected(data=attention_state.context,
+                                               num_hidden=self.num_hidden,
+                                               weight=self.mapped_context_w,
+                                               bias=self.mapped_context_b,
+                                               name="%smapped_context_fc_t%d" % (self.prefix, seq_idx))
+
+        hidden = gate * mapped_rnn_output + (1 - gate) * mapped_context
+
+        if self.config.hidden_dropout > 0:
```

Comment:
also here, should the dropout not be applied to the input rather than applying it before the activation?